src/stackdriver_agent.c

```diff
@@ -0,0 +1,223 @@
+    if (plugin_dispatch_values(&vl) != 0) {
+        ERROR("%s: plugin_dispatch_values failed.", this_plugin_name);
+        return -1;
+    }
+    return 0;
+}
+
+static int sagt_submit_gauge(const char *type_instance, const char *plugin_instance,
+    cdtime_t now, cdtime_t interval, gauge_t gauge, meta_data_t *meta_data)
+{
+    value_t v = {.gauge = gauge};
+    return sagt_submit_helper("gauge", type_instance, plugin_instance, now, interval, &v,
+        meta_data);
+}
+
+static int sagt_submit_derive(const char *type_instance, const char *plugin_instance,
+    cdtime_t now, cdtime_t interval, derive_t derive, meta_data_t *meta_data)
+{
+    value_t v = {.derive = derive};
+    return sagt_submit_helper("derive", type_instance, plugin_instance, now, interval, &v,
+        meta_data);
+}
+
+
+/**
+ * Send a variety of agent status/health-related metrics.
+ */
+static int sagt_read(user_data_t *user_data)
+{
+    context_t *ctx = user_data->data;
+    cdtime_t now = cdtime();
+    cdtime_t interval = plugin_get_interval();
+
+    // This value list exists merely for the purpose of harvesting its key fields for the purpose
+    // of looking stuff up in the cache.
+    value_list_t vl = {};  // zero-init
+    sstrncpy(vl.plugin, this_plugin_name, sizeof(vl.plugin));
+
+    // version
+    {
+        meta_data_t *md = meta_data_create();
+        if (meta_data_add_string(md, "version", COLLECTD_USERAGENT) == 0)
+        {
+            sagt_submit_gauge("version", NULL, now, interval, 1, md);
+        }
+        meta_data_destroy(md);
+    }
+
+    // uptime
+    {
+        derive_t uptime = CDTIME_T_TO_TIME_T(now - ctx->start_time);
+        sagt_submit_derive("uptime", NULL, now, interval, uptime, NULL);
+    }
+
+    // memory used
+    {
+        FILE *f = fopen("/proc/self/statm", "r");
+        if (f)
+        {
+            size_t vm = 0;
+            if (fscanf(f, "%zu", &vm))
+            {
+                long page_size = sysconf(_SC_PAGESIZE);
+                size_t mused = vm * page_size;
+                sagt_submit_gauge("memory_usage", NULL, now, interval, mused, NULL);
+            }
+            fclose(f);
+        }
+    }
+
+    // Stats for API requests. The corresponding uc_meta_data_set calls are in
+    // write_gcm.c.
+    {
+        uint64_t value;
+        if (uc_meta_data_get_unsigned_int(&vl, SAGT_API_REQUESTS_SUCCESS, &value) == 0)
+        {
+            sagt_submit_derive("api_request_count", "success", now, interval, value, NULL);
+        }
+        if (uc_meta_data_get_unsigned_int(&vl, SAGT_API_REQUESTS_CONNECTIVITY_FAILURES, &value) == 0)
+        {
+            sagt_submit_derive("api_request_count", "connectivity_failures", now, interval, value, NULL);
+        }
+        if (uc_meta_data_get_unsigned_int(&vl, SAGT_API_REQUESTS_ERRORS, &value) == 0)
+        {
+          sagt_submit_derive("api_request_count", "errors", now, interval, value, NULL);
+        }
+    }
+
+    // configured_plugins: TODO
```

Comment:
Eh, I think I'll delete it. It was meant to be a placeholder for some future work where we send information about configured plugins, but I don't think there's any need to leave a placeholder there.